src/readers/lsbi.c

```diff
@@ -0,0 +1,193 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <errno.h>
+#include <stdbool.h>
+
+#include <readers.h>
+#include <logs.h>
+
+#define PATTERN_SIZE 4
+#define EXTENSION_BLOCK_LENGTH 8
+#define LSB(x, n) ((x) & ((1 << (n)) - 1))
+
+
+Stego *retrieve_lsbi(FILE *file, size_t offset, char **extension)
+{
+    if (fseek(file, offset, SEEK_SET) < 0)
+    {
+        perror("Failed to seek to offset in file");
+        return NULL;
+    }
+
+    int pattern[PATTERN_SIZE] = {0};
+
+    uint8_t pattern_bytes[PATTERN_SIZE];
+    
+    if (fread(pattern_bytes, sizeof(uint8_t), PATTERN_SIZE, file) != PATTERN_SIZE)
+    {
+        perror("Not enough bytes to read pattern");
+        return NULL;
+    }
+
+    for(int i = 0 ; i < PATTERN_SIZE ; i++)
+    {
+        pattern[i] = pattern_bytes[i] & 1;
+        LOG("Pattern %d inversion: %d\n", i, pattern[i]);
+    }
+
+    /*
+                0  1  2  3  4  5 ...
+    B  G  R  B  G  R  B  G  R  B ...
+    00 01 10 11 S  -  I  Z  -  E ...
+    */
+
+    size_t bits_needed = sizeof(uint32_t) * 8;
+    size_t bytes_count = (bits_needed * 3 + 1) / 2;
+    uint8_t *length_bytes = malloc(bytes_count);
+    if (length_bytes == NULL)
+    {
+        perror("Memory allocation for length failed");
+        return NULL;
+    }
+
+    if (fread(length_bytes, sizeof(uint8_t), bytes_count, file) != bytes_count)
+    {
+        perror("Not enough bytes to read length");
+        return NULL;
+    }
+
+    uint32_t message_length = 0;
+    size_t bits_read = 0;
+    for (int i = 0 ; i < bytes_count && bits_read < bits_needed ; i++)
+    {
+        if (i % 3 != 1) // not R
+        {
+            uint8_t byte = length_bytes[i];
+            uint8_t last3 = LSB(byte, 3);
+            int pattern_index = last3 >> 1;
+            uint8_t bit = (last3 & 1) ^ pattern[pattern_index];
+
+            message_length = (message_length << 1) | (bit & 1);
+
+            bits_read++;
+        }
+    }
+    free(length_bytes);
+
+    if (bits_read < bits_needed)
+    {
+        LOG("Could not read enough bits for length: got %zu expected %zu\n", bits_read, bits_needed);
+        return NULL;
+    }
+
+    LOG("Message length: %u\n", message_length);
+
+    Stego *stego = malloc(sizeof(Stego) + message_length);
+    if (stego == NULL)
+    {
+        perror("Memory allocation for Stego failed");
+        return NULL;
+    }
+
+    size_t total_bits = (size_t)message_length * 8;
+    size_t bits_collected = 0;
+    int index = 0;
+
+    while (bits_collected < total_bits)
+    {
+        char byte = 0;
+        do {
+            byte = fgetc(file);
+            if (feof(file))
+            {
+                perror("Not enough bytes to read message");
+
+                free(stego);
+
+                return NULL;
+            }
+            index++;
+        } while ((index - 1) % 3 == 1); // skip R
```

Comment:
Nit

```suggestion
        } while (index++ % 3 == 1); // skip R
```

Can't tell which one looks cleaner 😂